database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift

```diff
@@ -47,9 +49,33 @@ struct Post: Identifiable {
     self.title = title
     self.body = body
     self.starCount = starCount
+    starDictionary = [:]
   }
 
   init() {
     self.init(id: "", uid: "", author: "", title: "", body: "")
   }
+
+  mutating func didTapStarButton(isStarred: Bool) {
+    let ref: DatabaseReference = {
+      Database.database().reference()
+    }()
+    if isStarred {
+      // add current user to the post.starDictionary
+      starDictionary[uid] = true
+      ref.child("/posts/\(id)/starDictionary").setValue(starDictionary)
+      ref.child("/user-posts/\(uid)/\(id)/starDictionary").setValue(starDictionary)
+      // increment starCount on current post
+      ref.child("/posts/\(id)/starCount").setValue(starCount + 1)
+      ref.child("/user-posts/\(uid)/\(id)/starCount").setValue(starCount + 1)
+    } else {
+      // remove current user from the post.starDictionary
+      starDictionary.removeValue(forKey: uid)
+      ref.child("/posts/\(id)/starDictionary").setValue(starDictionary)
+      ref.child("/user-posts/\(uid)/\(id)/starDictionary").setValue(starDictionary)
+      // decrement starCount on current post
+      ref.child("/posts/\(id)/starCount").setValue(starCount - 1)
+      ref.child("/user-posts/\(uid)/\(id)/starCount").setValue(starCount - 1)
```

Comment:
If you're going to go down this route, you should use a [transaction](https://firebase.google.com/docs/database/ios/read-and-write#save_data_as_transactions).